Cubical/Data/Maybe/Properties.agda

```diff
@@ -0,0 +1,105 @@
+{-# OPTIONS --cubical --safe #-}
+module Cubical.Data.Maybe.Properties where
+
+open import Cubical.Core.Everything
+open import Cubical.Foundations.Prelude
+open import Cubical.Foundations.HLevels
+open import Cubical.Foundations.Isomorphism
+open import Cubical.Data.Empty
+open import Cubical.Data.Unit
+open import Cubical.Data.Nat
+open import Cubical.Relation.Nullary
+open import Cubical.Data.Sum
+
+open import Cubical.Data.Maybe.Base
+
+-- Path space of Maybe type
+module MaybePath {ℓ} {A : Type ℓ} where
+  Cover : Maybe A → Maybe A → Type ℓ
+  Cover nothing  nothing   = Lift Unit
+  Cover nothing  (just _)  = Lift ⊥
+  Cover (just _) nothing   = Lift ⊥
+  Cover (just a) (just a') = a ≡ a'
+
+  reflCode : (c : Maybe A) → Cover c c
+  reflCode nothing  = lift tt
+  reflCode (just b) = refl
+
+  encode : ∀ c c' → c ≡ c' → Cover c c'
+  encode c _ = J (λ c' _ → Cover c c') (reflCode c)
+
+  encodeRefl : ∀ c → encode c c refl ≡ reflCode c
+  encodeRefl c = JRefl (λ c' _ → Cover c c') (reflCode c)
+  
+  decode : ∀ c c' → Cover c c' → c ≡ c'
+  decode nothing  nothing  _ = refl
+  decode (just _) (just _) p = cong just p
+
+  decodeRefl : ∀ c → decode c c (reflCode c) ≡ refl
+  decodeRefl nothing  = refl
+  decodeRefl (just _) = refl
+
+  decodeEncode : ∀ c c' → (p : c ≡ c') → decode c c' (encode c c' p) ≡ p
+  decodeEncode c _ =
+    J (λ c' p → decode c c' (encode c c' p) ≡ p)
+      (cong (decode c c) (encodeRefl c) ∙ decodeRefl c)
+
+  isOfHLevelCover : (n : ℕ)
+    → isOfHLevel (suc (suc n)) A
+    → ∀ c c' → isOfHLevel (suc n) (Cover c c')
+  isOfHLevelCover n p nothing  nothing   = isOfHLevelLift (suc n) (isOfHLevelUnit (suc n))
+  isOfHLevelCover n p nothing  (just a') = isOfHLevelLift (suc n)
+    (subst (λ m → isOfHLevel m ⊥) (+-comm n 1) (hLevelLift n isProp⊥))
+  isOfHLevelCover n p (just a) nothing   = isOfHLevelLift (suc n)
+    (subst (λ m → isOfHLevel m ⊥) (+-comm n 1) (hLevelLift n isProp⊥))
+  isOfHLevelCover n p (just a) (just a') = p a a'
+
+isOfHLevelMaybe : ∀ {ℓ} (n : ℕ) {A : Type ℓ}
+  → isOfHLevel (suc (suc n)) A
+  → isOfHLevel (suc (suc n)) (Maybe A)
+isOfHLevelMaybe n lA c c' =
+  retractIsOfHLevel (suc n)
+    (MaybePath.encode c c')
+    (MaybePath.decode c c')
+    (MaybePath.decodeEncode c c')
+    (MaybePath.isOfHLevelCover n lA c c')
+
+private
+ variable
+   ℓ : Level
+   A : Type ℓ
+
+fromJust-def : A → Maybe A → A
+fromJust-def a nothing = a
+fromJust-def _ (just a) = a
+
+just-inj : (x y : A) → just x ≡ just y → x ≡ y
+just-inj x _ eq = cong (fromJust-def x) eq 
+
+discreteMaybe : Discrete A → Discrete (Maybe A)
+discreteMaybe eqA nothing nothing    = yes refl
+discreteMaybe eqA nothing (just a')  = no (λ p → lower (subst (caseMaybe (Maybe _) (Lift ⊥)) p (just a')))
+discreteMaybe eqA (just a) nothing   = no (λ p → lower (subst (caseMaybe (Lift ⊥) (Maybe _)) p (just a)))
```

Comment:
I think this should be factored out into its own definition stating that just and nothing are never equal

Reply:
You're right, that lemma is useful. Done.